src/physics/em/detail/SBEnergyDistribution.hh

```diff
@@ -150,19 +150,26 @@ template<class X>
 template<class Engine>
 CELER_FUNCTION auto SBEnergyDistribution<X>::operator()(Engine& rng) -> Energy
 {
+    // Maximum number of iterations in rejection loop
+    constexpr size_type max_iter = 100;
     // Sampled energy
     Energy exit_energy;
     // Calculated cross section used inside rejection sampling
     real_type xs{};
-    do
+    for (CELER_MAYBE_UNUSED size_type i : range(max_iter))
     {
         // Sample scaled energy and subtract correction factor
         exit_energy = helper_.sample_exit_energy(rng);
 
         // Interpolate the differential cross setion at the sampled exit energy
         xs = helper_.calc_xs(exit_energy).value() * scale_xs_(exit_energy);
         CELER_ASSERT(xs >= 0 && xs <= 1 / inv_max_xs_);
-    } while (!BernoulliDistribution(xs * inv_max_xs_)(rng));
+
+        if (BernoulliDistribution(xs * inv_max_xs_)(rng))
+        {
+            break;
+        }
```

Comment:
I don't think we want this. The SB sampling algorithm in geant4 uses a bunch of heuristics to try to reduce the rejection rate, and the one I implemented is much more precisely able to reject, so I think the rejection rate should be much lower. I'd feel more comfortable with an assertion for a counter not exceeding 100, so we can see if/when it happens in practice (or some diagnostic or something). The rejection should only affect performance and not the results. Also, breaking early would potentially return an exit energy that's incorrect, right?

Reply:
Hmm ok, I actually only added this because with the positron correction fix I was running into some very long rejection loops (long enough that the code seemed to be hanging, which is the only reason I noticed). True, the distribution would be incorrect if we break early, but at least it would be incorrect like geant4 ;).

Reply:
Interesting. I would have thought that when Geant4 exited it would kill the track or raise an error or something... I saw the `G4Exception` code in the SB sampling and figured that was for the count limit but I guess not.

If the positron correction is hanging but the electron sampling isn't, there might be something else wrong. Is it with a particular energy range? Is it quantifiable using the diagnostic RNG engine?

Reply:
Is the max xs supposed to be scaled by the correction factor? I can't see where that's done in the G4 SB model.

Reply:
The max xs is used for rejection in `SBEnergyDistribution` as `helper.max_xs().value() * scale_xs(helper.max_xs_energy())`, but I see now that this assumes the maximum scaled cross section is at the same energy as the maximum original cross section -- which might not be the case if `scale_xs` is nonlinear. So it might be that the maximum scaled xs for positrons is at a different energy from electrons?

Reply:
(Copying from discussion with Amanda on slack)

- rather than finding the maximum cross section over all energies we should be finding the max xs above the cutoff energy
- our SB cross section tables need to be consistent with the model applicability and the cutoffs
- cutoffs are material dependent, not Z dependent though.. so with our current approach would we have to find the max for each Z for each material?
- per positron/electron as well, assuming the scaling changes where the max is
- the 'max' cross section is currently underestimated by a ridiculous amount at the lower end (see plot below)
- the positron scaling is so strong that we can always set the max to the first energy point -- it's exponential

![copper-sb-table](https://user-images.githubusercontent.com/741229/162484768-971b59be-0756-4660-b614-e970ffc257c8.png)
![copper-sb-slices](https://user-images.githubusercontent.com/741229/162484771-6822bf23-81ba-4383-af20-5ca4189e93cc.png)

@amandalund is currently updating the `argmax` to be dependent per material, Z, and particle type; and updating the XsScaling helper classes accordingly since the argmax depends on the scaling.

Reply:
After looking at this again I decided I don't think I was right about needing to get the max xs above the cutoff energy.. it could _maybe_ reduce the rejection rate in some cases, but it shouldn't affect the results. So I've kept `argmax` the same and made sure we're using the correct positron max xs (at the first reduced photon energy grid point).